evm-adapters/src/sputnik/forked_backend/rpc.rs

```diff
@@ -35,6 +32,79 @@ pub struct ForkMemoryBackend<B, M> {
     pin_block_meta: Block<TxHash>,
     /// The chain id of the forked chain
     chain_id: U256,
+    /// The cache file to store any state.
+    /// - Some(Some("path")) -> it gets stored at "path"
+    /// - Some(None) -> it gets stored at ~/.foundry/cache/{pin_block}.json
+    /// - None -> it does not get stored
+    cache_file: Option<Option<PathBuf>>,
+}
+
+#[derive(thiserror::Error, Debug)]
+enum ForkError {
+    #[error(transparent)]
+    Serde(#[from] serde_json::Error),
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+    #[error("no cache file was specified")]
+    NoCache,
+}
+
+fn dump_cache(path: PathBuf, cache: &BTreeMap<H160, MemoryAccount>) -> Result<(), ForkError> {
+    std::fs::create_dir_all(path.parent().expect("no parent found"))?;
+    let file = std::fs::File::create(path)?;
+    let file = std::io::BufWriter::new(file);
+    serde_json::to_writer(file, cache)?;
+    Ok(())
+}
+
+fn load_cache(
+    path: Option<Option<PathBuf>>,
+    pin_block: Option<u64>,
+) -> Result<BTreeMap<H160, MemoryAccount>, ForkError> {
+    if let Some(pin_block) = pin_block {
+        let path = match path {
+            Some(Some(path)) => path,
+            Some(None) => {
+                let path = dirs_next::home_dir()
+                    .map(|p| p.join(".foundry").join("cache").join(pin_block.to_string()))
+                    .expect("could not construct pin block path");
+                path
+            }
+            None => return Err(ForkError::NoCache),
+        };
+
+        let file = std::fs::File::open(path)?;
+        let file = std::io::BufReader::new(file);
+        Ok(serde_json::from_reader(file)?)
+    } else {
+        Err(ForkError::NoCache)
+    }
+}
+
+impl<B, M> Drop for ForkMemoryBackend<B, M> {
+    fn drop(&mut self) {
+        let pin_block = match self.pin_block.expect("no pin block found; this should never happen")
+        {
+            BlockId::Number(BlockNumber::Number(num)) => num,
+            _ => panic!("non block number pin blocks not supported"),
```

Comment:
this seems (guaranteed) to never panic?
otherwise we should simply skip dumping because worst case, everything is finished but we panic on program shutdown